distribution/lib/Standard/Database/0.0.0-dev/src/Data/Table.enso

```diff
@@ -681,100 +650,103 @@ type Table
+
+       ? Joining on equality of columns with the same name
+
+         When joining two columns with the same name and an equality condition,
+         only one copy of column will be included in the output (avoiding
+         unnecessary duplication and renaming).
+
+       ? Same-name column join shorthand
+
+         As a shorthand, providing a column name or a list of column names
+         allows to join the two tables on equality of corresponding columns with
+         the same name. So `table.join other on=["A", "B"]` is a shorthand for:
+             table.join other on=[Join_Condition.Equals "A" "A", Join_Condition.Equals "B" "B"]
+    join : Table -> Join_Kind -> Join_Condition | Text | Vector (Join_Condition | Text) -> Text -> Problem_Behavior -> Table
+    join self right join_kind=Join_Kind.Inner on=[Join_Condition.Equals 0 0] right_prefix="Right_" on_problems=Report_Warning =
+        can_proceed = if Table_Helpers.is_table right . not then Error.throw (Type_Error.Error Table right "right") else
+            same_backend = case right of
+                            _ : Table -> True
+                            _ -> False
+            if same_backend . not then Error.throw (Illegal_Argument.Error "Currently cross-backend joins are not supported. You need to upload the table in-memory table before joining it with a database one, or materialize this table.") else
+                True
+        if can_proceed then
+            left = self
+            new_table_name = left.name + "_" + right.name
+
+            needed_indicators = case join_kind of
+                Join_Kind.Left_Exclusive  -> Pair.new False True
+                Join_Kind.Right_Exclusive -> Pair.new True False
+                _ -> Pair.new False False
+            subquery_setups = Database_Join_Helper.prepare_subqueries left right needed_indicators.first needed_indicators.second
+            left_setup = subquery_setups.first
+            right_setup = subquery_setups.second
+
+            join_resolution = Database_Join_Helper.make_join_helpers left right left_setup.column_mapping right_setup.column_mapping . resolve on on_problems
+            result_columns = Database_Join_Helper.select_columns_for_join join_kind left_setup.new_columns right_setup.new_columns join_resolution.redundant_column_names right_prefix
+
+            # TODO proper equality of nulls in join conditions
+            on_expressions = join_resolution.conditions
+
+            where_expressions = case join_kind of
+                Join_Kind.Left_Exclusive  ->
+                    is_right_missing = SQL_Expression.Operation "IS_NULL" [right_setup.indicator_column.expression]
+                    [is_right_missing]
+                Join_Kind.Right_Exclusive ->
+                    is_left_missing = SQL_Expression.Operation "IS_NULL" [left_setup.indicator_column.expression]
+                    [is_left_missing]
+                _ -> []
+
+            internal_kind = case join_kind of
```

Comment:
shouldn't this conversion be part of (private) API on Join_Kind?

Reply:
You mean on the `Table.Join_Kind` or the `Internal_Join_Kind`? (Sorry for confusion, as James rightfully noted, I need to make these more distinguishable.)

The (Table)`Join_Kind` is part of the backend-agnostic common Table API, while this conversion is strictly part of the Database backends implementation. Even more so, it is defined in the `Table` library which does not necessarily need to know that `Database` exists at all.

If it was the other way around - yeah I guess we can move it, I'll see if it makes sense.

Reply:
Ok, I tried implementing it and after reminding more details of this code I think it should stay more or less as-is. This particular conversion is not really quite 1-1 and it is an implementation detail of how we are approaching the joins here. So the thing is, both `Left_Exclusive` and `Left_Outer` are translated to a `Left` join, with some differing characteristics. So I'd keep this logic close to the join itself. However, if you still think I should move it, I guess I can. I'm not 100% convinced this is the only right way.